### PR TITLE
Allow SGLang to still work if Triton is not installed.

### PR DESCRIPTION
**Pull request by jonahbernard**
## Motivation

Prevents SGLang from immediately crashing if Triton is not installed. This is useful for running on hardware that does not use Triton. Adapted from https://github.com/vllm-project/vllm/pull/6786.

## Modifications

Creates a mock Triton library that is used if Triton is not available in the environment. 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

**Comment by Jonahcb** [comment]
@ping1jing2 do you want me implement the second sglang-bot suggestion (lint checker)? 

The second one would be useful, but I don't know what the SGLang team's rules are for the lint checker so I want to confirm that you want me to implement the lint check.

File: python/sglang/test/test_cutlass_moe.py [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 import torch
-import triton  # Added import
-import triton.testing  # Added import
 from transformers import AutoConfig
 
 from sglang.srt.layers.moe.cutlass_moe import cutlass_fused_experts_fp8
 from sglang.srt.layers.moe.fused_moe_triton.fused_moe import fused_experts
 from sglang.srt.layers.moe.moe_runner.base import MoeRunnerConfig
 from sglang.srt.layers.moe.topk import StandardTopKOutput
+from sglang.srt.triton_utils import triton  # Added import
```
Comment (ping1jing2) [inline python/sglang/test/test_cutlass_moe.py:10]: maybe you missed the `testing` here

Comment (jonahbernard) [reply python/sglang/test/test_cutlass_moe.py:10]: @ping1jing2 thank you for looking! `testing` was removed by the `pre-commit` because `testing` is accessed from the imported `triton'. `pre-commit` removed the `testing` in a lot of other cases as well.

File: scripts/lint/check_direct_triton_imports.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+]
+
+
+def main() -> int:
+    errors: list[str] = []
+
+    for root in SCAN_ROOTS:
+        root_path = pathlib.Path(root)
+        if not root_path.exists():
+            continue
```
Comment (ping1jing2) [inline scripts/lint/check_direct_triton_imports.py]: 1. These directories must always exist. If not, we should throw a warning or error msg rather than continue.
2. maybe we should use `if not root_path.is_dir():` here

File: scripts/lint/check_direct_triton_imports.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            if any(str(p).startswith(a) for a in ALLOWED):
+                continue
+            try:
+                source = p.read_text(encoding="utf-8", errors="replace")
+            except OSError:
+                continue
+            try:
+                tree = ast.parse(source, filename=str(p))
+            except SyntaxError:
+                continue
```
Comment (ping1jing2) [inline scripts/lint/check_direct_triton_imports.py:71]: 1. we may use `p.read_byte()` here due to performance considerations, because I've checked that `source` can either be a normal string, a byte string, or an AST object in `ast.parse`
2. hope you can combine these two `try-except` into one.

File: scripts/lint/check_direct_triton_imports.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                        if alias.name == "triton" or alias.name.startswith("triton."):
+                            errors.append(
+                                f"{p}:{node.lineno}: direct `import {alias.name}` found"
+                            )
+                elif isinstance(node, ast.ImportFrom) and node.module and (
+                    node.module == "triton" or node.module.startswith("triton.")
+                ):
+                    errors.append(
+                        f"{p}:{node.lineno}: direct `from {node.module} import ...` found"
+                    )
```
Comment (ping1jing2) [inline scripts/lint/check_direct_triton_imports.py]: Actually, I am not faimilar with `AST` but I found `ast.NodeVisitor` class, and Gemini told me we can optimize it like the code below, please correct me if i am wrong.

```python
class TritonImportChecker(ast.NodeVisitor):
    def __init__(self, filename):
        self.filename = filename
        self.errors = []
        self._in_try_block = 0

    def visit_Try(self, node):
        # 进入 Try 块，标记内部导入为受保护
        self._in_try_block += 1
        self.generic_visit(node)
        self._in_try_block -= 1

    def _check_triton(self, name, lineno, is_from=False):
        if self._in_try_block > 0:
            return
        
        if name == "triton" or name.startswith("triton."):
            msg_type = f"from {name} import ..." if is_from else f"import {name}"
            self.errors.append(f"{self.filename}:{lineno}: direct `{msg_type}` found")

    def visit_Import(self, node):
        for alias in node.names:
            self._check_triton(alias.name, node.lineno)

    def visit_ImportFrom(self, node):
        if node.module:
            self._check_triton(node.module, node.lineno, is_from=True)


checker = TritonImportChecker(path_str)
checker.visit(tree)
all_errors.extend(checker.errors)
```

Comment (jonahbernard) [reply scripts/lint/check_direct_triton_imports.py]: @ping1jing2 thank you. this is a good improvement on the speed. done!

File: scripts/lint/check_direct_triton_imports.py (hunk 3 above)
Comment (jonahbernard) [reply scripts/lint/check_direct_triton_imports.py:71]: @ping1jing2 done

File: scripts/lint/check_direct_triton_imports.py (hunk 2 above)
Comment (jonahbernard) [reply scripts/lint/check_direct_triton_imports.py]: @ping1jing2 done, thank you

**Review by ping1jing2 — APPROVED** [review]
